File: src/main/java/no/nav/veilarbaktivitet/oppfolging/periode/OppfolgingsperiodeConsumer.java
package no.nav.veilarbaktivitet.oppfolging.periode;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.json.JsonUtils;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
class OppfolgingsperiodeConsumer {
    private final OppfolgingsperiodeService oppfolgingsperiodeService;

    @KafkaListener(topics = "${topic.inn.oppfolgingsperiode}", containerFactory = "stringStringKafkaListenerContainerFactory")
    void opprettEllerOppdaterSistePeriode(ConsumerRecord<String, String> consumerRecord) {
        SisteOppfolgingsperiodeV1 sisteOppfolgingsperiodeV1 = JsonUtils.fromJson(consumerRecord.value(), SisteOppfolgingsperiodeV1.class);
        log.info("oppfølgingsperiode: {}", sisteOppfolgingsperiodeV1);
        oppfolgingsperiodeService.upsertOppfolgingsperiode(sisteOppfolgingsperiodeV1);

        if(sisteOppfolgingsperiodeV1.sluttDato != null) {
            oppfolgingsperiodeService.avsluttOppfolgingsperiode(sisteOppfolgingsperiodeV1.uuid, sisteOppfolgingsperiodeV1.sluttDato);
        }


    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/oppfolging/periode/OppfolgingsperiodeConsumer.kt
package no.nav.veilarbaktivitet.oppfolging.periode

import no.nav.common.json.JsonUtils
import no.nav.common.kafka.consumer.ConsumeStatus
import no.nav.common.kafka.consumer.TopicConsumer
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class OppfolgingsperiodeConsumer(
    private val oppfolgingsperiodeService: OppfolgingsperiodeService
) : TopicConsumer<String, String> {
    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    override fun consume(consumerRecord: ConsumerRecord<String?, String?>): ConsumeStatus {
        val sisteOppfolgingsperiodeV1 = JsonUtils.fromJson(
            consumerRecord.value(),
            SisteOppfolgingsperiodeV1::class.java
        )
        log.info("oppfølgingsperiode: {}", sisteOppfolgingsperiodeV1)
        oppfolgingsperiodeService.upsertOppfolgingsperiode(sisteOppfolgingsperiodeV1)
        if (sisteOppfolgingsperiodeV1.sluttDato != null) {
            oppfolgingsperiodeService.avsluttOppfolgingsperiode(
                sisteOppfolgingsperiodeV1.uuid,
                sisteOppfolgingsperiodeV1.sluttDato
            )
        }
        return ConsumeStatus.OK
    }
}
